File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldHaveCorrectSignature.NUnit.vb
Imports NUnit.Framework

Namespace Tests.Diagnostics
    Class NUnitTest
        <NUnit.Framework.Test>
        Private Sub PrivateTestMethod()
        End Sub

        <Test>
        Protected Sub ProtectedTestMethod()
        End Sub

        <Test>
        Friend Sub InternalTestMethod()
        End Sub

        <Test>
        Public Async Sub AsyncTestMethod()
        End Sub

        <Test>
        Public Sub GenericTestMethod(Of T)()
        End Sub

        <Test>
        Public Sub ValidTest()
        End Sub
    End Class

    Class NUnitTest_TestCase
        <NUnit.Framework.TestCase(42)>
        Private Sub PrivateTestMethod(ByVal data As Integer)
        End Sub

        <TestCase(42)>
        Protected Sub ProtectedTestMethod(ByVal data As Integer)
        End Sub

        <TestCase(42)>
        Friend Sub InternalTestMethod(ByVal data As Integer)
        End Sub

        <TestCase(42)>
        Public Async Sub AsyncTestMethod(ByVal data As Integer)
        End Sub

        <TestCase(42)>
        Public Sub GenericTestMethod(Of T)(ByVal data As T)
        End Sub

        <TestCase(42)>
        Public Sub ValidTest(ByVal data As Integer)
        End Sub
    End Class

    Class NUnitTest_TestCaseSource
        Public Shared DataProvider As Object() = {42}

        <NUnit.Framework.TestCaseSource("DataProvider")>
        Private Sub PrivateTestMethod(ByVal data As Object)
        End Sub

        <TestCaseSource("DataProvider")>
        Protected Sub ProtectedTestMethod(ByVal data As Object)
        End Sub

        <TestCaseSource("DataProvider")>
        Friend Sub InternalTestMethod(ByVal data As Object)
        End Sub

        <TestCaseSource("DataProvider")>
        Public Async Sub AsyncTestMethod(ByVal data As Object)
        End Sub

        <TestCaseSource("DataProvider")>
        Public Sub GenericTestMethod(Of T)(ByVal data As T)
        End Sub

        <TestCaseSource("DataProvider")>
        Public Sub ValidTest(ByVal data As Object)
        End Sub
    End Class

    Public Class NUnitTest_Theories
        <NUnit.Framework.Theory>
        Private Sub PrivateTestMethod(ByVal data As Integer)
        End Sub

        <Theory>
        Protected Sub ProtectedTestMethod(ByVal data As Integer)
        End Sub

        <Theory>
        Friend Sub InternalTestMethod(ByVal data As Integer)
        End Sub

        <Theory>
        Public Async Sub AsyncTestMethod(ByVal data As Integer)
        End Sub

        <Theory>
        Public Sub GenericTestMethod(Of T)(ByVal data As T())
        End Sub

        <Theory>
        Public Sub ValidTest(ByVal data As Integer)
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/TestMethodShouldHaveCorrectSignature.NUnit.cs
using NUnit.Framework;

namespace Tests.Diagnostics
{
    class NUnitTest
    {
        [NUnit.Framework.Test]
        private void PrivateTestMethod() // Noncompliant
        {
        }

        [Test]
        protected void ProtectedTestMethod() // Noncompliant
        {
        }

        [Test]
        internal void InternalTestMethod() // Noncompliant
        {
        }

        [Test]
        public async void AsyncTestMethod()  // Noncompliant
        {
        }

        [Test]
        public void GenericTestMethod<T>()  // Noncompliant
        {
        }

        [Test]
        public void ValidTest()
        {
        }
    }

    class NUnitTest_TestCase
    {
        [NUnit.Framework.TestCase(42)]
        private void PrivateTestMethod(int data) // Noncompliant
        {
        }

        [TestCase(42)]
        protected void ProtectedTestMethod(int data) // Noncompliant
        {
        }

        [TestCase(42)]
        internal void InternalTestMethod(int data) // Noncompliant
        {
        }

        [TestCase(42)]
        public async void AsyncTestMethod(int data)  // Noncompliant
        {
        }

        [TestCase(42)]
        public void GenericTestMethod<T>(T data)  // valid
        {
        }

        [TestCase(42)]
        public void ValidTest(int data)
        {
        }
    }

    class NUnitTest_TestCaseSource
    {
        public static object[] DataProvider = { 42 };

        [NUnit.Framework.TestCaseSource("DataProvider")]
        private void PrivateTestMethod(object data) // Noncompliant
        {
        }

        [TestCaseSource("DataProvider")]
        protected void ProtectedTestMethod(object data) // Noncompliant
        {
        }

        [TestCaseSource("DataProvider")]
        internal void InternalTestMethod(object data) // Noncompliant
        {
        }

        [TestCaseSource("DataProvider")]
        public async void AsyncTestMethod(object data)  // Noncompliant
        {
        }

        [TestCaseSource("DataProvider")]
        public void GenericTestMethod<T>(T data)  // compliant
        {
        }

        [TestCaseSource("DataProvider")]
        public void ValidTest(object data)
        {

        }
    }

    public class NUnitTest_Theories
    {
        [NUnit.Framework.Theory]
        private void PrivateTestMethod(int data) // Noncompliant
        {
        }

        [Theory]
        protected void ProtectedTestMethod(int data) // Noncompliant
        {
        }

        [Theory]
        internal void InternalTestMethod(int data) // Noncompliant
        {
        }

        [Theory]
        public async void AsyncTestMethod(int data)  // Noncompliant
        {
        }

        [Theory]
        public void GenericTestMethod<T>(T[] data)  // Noncompliant
        {
        }

        [Theory]
        public void ValidTest(int data)
        {
        }
    }
}
